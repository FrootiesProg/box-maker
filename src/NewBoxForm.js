import React, { useState } from "react";

// This is a functional component named "NewBoxForm" that allows users to input box data and add a new box.
function NewBoxForm({ addBox }) {
  const initialState = { color: "", width: "", height: "" };
  const [formData, setFormData] = useState(initialState); // State to store form data.

  // Event handler to update form data as user types.
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((prevData) => ({
      ...prevData,
      [name]: value, // Update the specific field (color, width, or height) in form data.
    }));
  };

  // Event handler for form submission.
  const handleSubmit = (e) => {
    e.preventDefault(); // Prevent the default form submission behavior.
    addBox(formData); // Call the "addBox" function passed as a prop, passing in the form data.
    setFormData(initialState); // Reset the form data to initial state after adding a box.
  };

  // Return JSX to render the NewBoxForm component.
  return (
    <form onSubmit={handleSubmit}>
      {/* Input field for color */}
      <label htmlFor="color">Color:</label>
      <input
        type="text"
        id="color"
        name="color"
        value={formData.color}
        onChange={handleChange}
      />

      {/* Input field for width */}
      <label htmlFor="width">Width:</label>
      <input
        type="text"
        id="width"
        name="width"
        value={formData.width}
        onChange={handleChange}
      />

      {/* Input field for height */}
      <label htmlFor="height">Height:</label>
      <input
        type="text"
        id="height"
        name="height"
        value={formData.height}
        onChange={handleChange}
      />

      {/* Button to submit the form */}
      <button>Add Box</button>
    </form>
  );
}

// Export the NewBoxForm component for use in other parts of the application.
export default NewBoxForm;
