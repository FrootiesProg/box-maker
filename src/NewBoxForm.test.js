import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

// Test to ensure that the NewBoxForm component renders correctly and its elements are present.
test("renders NewBoxForm component", () => {
  // Render the NewBoxForm component and capture a snapshot of the rendered component.
  const { asFragment } = render(<NewBoxForm />);

  // Find form elements in the rendered component.
  const colorInput = screen.getByLabelText("Color:");
  const widthInput = screen.getByLabelText("Width:");
  const heightInput = screen.getByLabelText("Height:");
  const addButton = screen.getByText("Add Box");

  // Expect form elements to be present in the DOM.
  expect(colorInput).toBeInTheDocument();
  expect(widthInput).toBeInTheDocument();
  expect(heightInput).toBeInTheDocument();
  expect(addButton).toBeInTheDocument();

  // Capture a snapshot of the rendered component for future reference.
  expect(asFragment()).toMatchSnapshot();
});

// Test to simulate form submission and verify that a new box is added.
test("submitting form adds a new box", () => {
  // Create a mock function to simulate the "addBox" function.
  const addBoxMock = jest.fn();

  // Render the NewBoxForm component with the mock function.
  const { asFragment } = render(<NewBoxForm addBox={addBoxMock} />);

  // Find form elements in the rendered component.
  const colorInput = screen.getByLabelText("Color:");
  const widthInput = screen.getByLabelText("Width:");
  const heightInput = screen.getByLabelText("Height:");
  const addButton = screen.getByText("Add Box");

  // Simulate user input by firing change events on input fields.
  fireEvent.change(colorInput, { target: { value: "blue" } });
  fireEvent.change(widthInput, { target: { value: "100" } });
  fireEvent.change(heightInput, { target: { value: "100" } });

  // Simulate form submission by clicking the "Add Box" button.
  fireEvent.click(addButton);

  // Expect the mock function to have been called with the specified arguments.
  expect(addBoxMock).toHaveBeenCalledWith({
    color: "blue",
    width: "100",
    height: "100",
  });

  // Capture a snapshot of the rendered component after form submission.
  expect(asFragment()).toMatchSnapshot();
});
