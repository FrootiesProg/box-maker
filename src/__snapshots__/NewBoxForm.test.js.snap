// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders NewBoxForm component 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="text"
      value=""
    />
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;

exports[`submitting form adds a new box 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="text"
      value=""
    />
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
